botorch_community/models/vbll_helper.py

```diff
@@ -0,0 +1,487 @@
+        prod = diag_term + factor_term
+        return prod.squeeze(-1) if reduce_dim else prod
+
+    def precision_weighted_inner_prod(self, b, reduce_dim=True):
+        raise NotImplementedError()
+
+    def __matmul__(self, inp):
+        assert inp.shape[-2] == self.loc.shape[-1]
+        assert inp.shape[-1] == 1
+        new_cov = self.covariance_weighted_inner_prod(
+            inp.unsqueeze(-3), reduce_dim=False
+        )
+        return Normal(self.loc @ inp, torch.sqrt(torch.clip(new_cov, min=1e-12)))
+
+    def squeeze(self, idx):
+        return LowRankNormal(
+            self.loc.squeeze(idx),
+            self.cov_factor.squeeze(idx),
+            self.cov_diag.squeeze(idx),
+        )
+
+
+class DenseNormalPrec(torch.distributions.MultivariateNormal):
+    """A DenseNormal parameterized by the mean and the cholesky decomp of the precision matrix.
+
+    This function also includes a recursive_update function which performs a recursive
+    linear regression update with effecient cholesky factor updates.
+    """
+
+    def __init__(self, loc, cholesky, validate_args=False):
+        prec = cholesky @ tp(cholesky)
+        super(DenseNormalPrec, self).__init__(
+            loc, precision_matrix=prec, validate_args=validate_args
+        )
+        self.tril = cholesky
+
+    @property
+    def mean(self):
+        return self.loc
+
+    @property
+    def chol_covariance(self):
+        raise NotImplementedError()
+
+    @property
+    def covariance(self):
+        warnings.warn(
+            "Direct matrix inverse for dense covariances is O(N^3), consider using eg inverse weighted inner product"
+        )
+        return cholesky_inverse(self.tril)
+
+    @property
+    def inverse_covariance(self):
+        return self.precision_matrix
+
+    @property
+    def logdet_covariance(self):
+        return -2.0 * torch.diagonal(self.tril, dim1=-2, dim2=-1).log().sum(-1)
+
+    @property
+    def trace_covariance(self):
+        return (
+            (torch.inverse(self.tril) ** 2).sum(-1).sum(-1)
+        )  # compute as frob norm squared
+
+    def covariance_weighted_inner_prod(self, b, reduce_dim=True):
+        assert b.shape[-1] == 1
+        prod = (torch.linalg.solve(self.tril, b) ** 2).sum(-2)
+        return prod.squeeze(-1) if reduce_dim else prod
+
+    def precision_weighted_inner_prod(self, b, reduce_dim=True):
+        assert b.shape[-1] == 1
+        prod = ((tp(self.tril) @ b) ** 2).sum(-2)
+        return prod.squeeze(-1) if reduce_dim else prod
+
+    def __matmul__(self, inp):
+        assert inp.shape[-2] == self.loc.shape[-1]
+        assert inp.shape[-1] == 1
+        new_cov = self.covariance_weighted_inner_prod(
+            inp.unsqueeze(-3), reduce_dim=False
+        )
+        return Normal(self.loc @ inp, torch.sqrt(torch.clip(new_cov, min=1e-12)))
+
+    def squeeze(self, idx):
+        return DenseNormalPrec(self.loc.squeeze(idx), self.tril.squeeze(idx))
+
+
+cov_param_dict = {
```

Comment:
Can you move this up to the top of the module and capitalize this to indicate this is a constant?